Fix library selection, save header configuration per crate/playlist

poelzi opened the pull request:
This branch fixes multiple selection issues in the library.

- When switching between library feature, the table may show a entry selected but currentIndex will not reflect the selection anymore. Move selection now actually queries the model selection and uses those row values for calculating the next entry.
- Wrap around when using the move COs. There is no benefit on being stuck at the end/top of the list.
- Use a signal to initiate a state save on the view.
- Use a key provided by the model for associating the model index with a scrollposition. This way, a model can be reused for different views and trigger a state save.

_____
edit (ronso0) fixes
#9547
#7984

poelzi committed:
Query selectionModel for selected rows

currentIndex is not necesarry in sync with the selected rows the user
sees. Query the selectionModel and using it results fixes those problems.

Also implement wrap-around behaviour.

Example:
Go to Library Tracks, select track. Go to AutoDJ select track,
go back to library. Now use moveSelection controlls will always
jump back to the first row because currentIndex returns invalid index.

The new implementation fixes this bug.

poelzi committed:
Use signal to initiate state save

Use a signal in Library or LibraryFeature objects to
save the scroll position of the trackmodel currently shown.
The key to store the scroll position is no longer defined by the
model pointer but a value provided by the model itself.
This way, the same model can be reused for different views like
the crates feature does.

ronso0 commented:
> Wrap around when using the move COs. There is no benefit on being stuck at the end/top of the list.

I'm not sure abou this. Sure with `[Libary],MoveVertical` you'll notice soon enough that you approach the end of the table.
But with `[Libary],ScrollVertical` wrapping will cause irritation IMO.

ronso0 commented:
Oh, I see this only affects the `Select[Prev/Next]Track` controls.

ronso0 commented on src/widget/wlibrarytableview.cpp:
```diff
@@ -92,36 +92,59 @@ void WLibraryTableView::moveSelection(int delta) {
-                selectRow(row + 1);
+        QItemSelectionModel* currentSelection = selectionModel();
+        if (currentSelection->selectedRows().length() > 0) {
+            if(delta > 0) {
```
```suggestion
            if (delta > 0) {
```

ronso0 commented on src/widget/wlibrarytableview.cpp:
```diff
@@ -92,36 +92,59 @@ void WLibraryTableView::moveSelection(int delta) {
+                if (row + 1 < pModel->rowCount()) {
+                    selectRow(row + 1);
+                } else {
+                    selectRow(0);
```
please add a wrap-around comment here

ronso0 commented on src/widget/wlibrarytableview.cpp:
```diff
@@ -92,36 +92,59 @@ void WLibraryTableView::moveSelection(int delta) {
     if (m_vScrollBarPosValues.contains(key)){
+        qDebug() << "restore key found:" << key << " pos: " << m_vScrollBarPosValues[key];
         verticalScrollBar()->setValue(m_vScrollBarPosValues[key]);
     }else{
```
```suggestion
    } else {
```
wouldn't pre-commit clang fix that?

ronso0 commented on src/widget/wlibrarytableview.cpp:
```diff
@@ -92,36 +92,59 @@ void WLibraryTableView::moveSelection(int delta) {
+    QString key = model->key();
+    qDebug() << "m_vScrollBarPosValues:" << m_vScrollBarPosValues;
 
     if (m_vScrollBarPosValues.contains(key)){
```
```suggestion
    if (m_vScrollBarPosValues.contains(key)) {
```

ronso0 commented:
> Oh, I see this only affects the `Select[Prev/Next]Track` controls.

..which don't require the table to be `:focus`ed. So it doesn't replace my fix for keyboard and `..MoveVertical` controls in #3049 , and afaict can tell it doesn't conflict either, right?

ronso0 commented:
as mentioned earlier (elsewher) It would be nice if  `[Library], MoveVertical ` would use moveSelection() (instead of key press emulation) if the tracks table is focused / was the last focused library widget. That way both Playlist,SelectTrack and Library, MoveVertical would benefit from the fixes introduced here.

poelzi committed:
Add save/restore functionality for library views

* Allow LibraryViews to save/restore their state like
scroll position, selection and currentIndex.
* Add state cache to baseplaylist
* Add special modelKey implementations for Playlists, Crates, Computer, Proxy
* Selection is now search aware and saved different pointers for different searches
* Clear Cache regularly

poelzi commented:
@ronso0 would like you give this version a test ?
@uklotzde is this still a candidate for 2.3 after some polishing or should I rebase on main ?

Unfortunately the small changes from the beginning got a bit larger but otherwise I could not implement it in a generic and working way for all the features. Now the selection of tracks, scrollposition and current Index is popery conserved between switches of different subsystems. Each playlist/crate/folder + search query has it's own state and will be garbage collected after LRU schema. (combined with the search combobox this is top).

Be-ing commented:
Did you rebase on 2.3? I don't see the recently added GitHub Actions checks.

ronso0 commented:
I guess I can test that in a few days with my controller, latest by the end of the week.
If it reliably saves the scroll position and selection for each feature this is a huge UX improvement and should definitely go into 2.3! Even more so if the navigation improvements also work for [Lib],Move.. controls (or if those can be mapped to the [Playlist] navi controls internally).

poelzi committed:
Add state save/restore to recordings/analyisis/missing features

poelzi committed:
Save/Restore support for externalplaylistmodel

poelzi commented:
Analyze/Recordings/Missing Tracks now work as well.
I would not test externalplaylist based features because rythmbox feature crashes here and I don't have any other tools to test.

Holzhaus commented:
> Analyze/Recordings/Missing Tracks now work as well.
> I would not test externalplaylist based features because rythmbox feature crashes here and I don't have any other tools to test.

Serato library can be tested by unpacking this zip onto a USB drive: https://github.com/mixxxdj/mixxx/pull/2480#issuecomment-582899196

ronso0 commented:
Just did a quick test with the keyboard: track selection and position are perfectly restored in internal lib features, also when removing sidebar items.

Holzhaus reviewed (commented):
There are some clazy and pre-commit issues.

poelzi commented:
@Holzhaus I know, there are also some debug statements I will remove before final review, but want to have some testing by more people first

JoergAtGithub commented:
I did a short test on Windows7, and it worked great!

poelzi committed:
Check TreeItem data before building

poelzi committed:
cleanup state safe/restoration code

poelzi commented:
@Holzhaus this one is ready for review/merge. I was able to check and serato and rythmbox after uwes patch and they work great.

poelzi commented:
> Did you rebase on 2.3? I don't see the recently added GitHub Actions checks.

It seems that most checks are only run on PR requests, not on Draft Pull Requests or, at least not when they where started as such. I think the default "on:" condition does not include that

ronso0 commented:
Close / Re-open ?

uklotzde commented on src/library/crate/cratetablemodel.cpp:
```diff
@@ -191,3 +191,14 @@ void CrateTableModel::removeTracks(const QModelIndexList& indices) {
+    if (this->m_selectedCrate.isValid()) {
+        return QString("0x%1/%2#%3")
+                .arg((quintptr)this, 0, 16, QChar('0'))
+                .arg((quintptr)this->m_selectedCrate.value())
```
Wrong cast

uklotzde commented:
I would prefer to add this feature to main instead of 2.3. Other opinions?

poelzi commented:
I'm mixed about this, it became quite some code that is new for 2.3. On the other hand, it fixes broken functionality and grearly improves useabiltiy. There where dozens of times I swared because i switched features and the list jumped back to the beginning when using a controller to navigate. So I think it would be nice to have the expected behaviour in 2.3

ronso0 commented:
Indeed, that scroll reset with controller navigation when switching features was actually _the_ UX annoyance for for me over the last years. If we can get that into 2.3 that would be awesome!
@uklotzde what are your concerns? Can we eliminate them with extensive testing?

uklotzde commented:
I'm fine with targeting 2.3 if this branch is tested thoroughly by others **before merging**. @poelzi This also requires your commitment to fix any bugs asap that might be discovered **after merging**!

But first let's fix the remaining issues.

poelzi committed:
Cleanup item selection code

Use stringbuilder.
use function to format pointer
more const

poelzi commented on src/library/crate/cratefeature.cpp:
```diff
@@ -817,5 +819,5 @@ void CrateFeature::slotTrackSelected(TrackPointer pTrack) {
 
 void CrateFeature::slotResetSelectedTrack() {
-    slotTrackSelected(TrackPointer());
+    //slotTrackSelected(TrackPointer());
```
I think this slipped in and I did not see it when I controlled the patch. Fixed.

poelzi commented:
@uklotzde all done

ronso0 commented:
Tested with a controller and [Library],MoveUp/Down and ..Scrollvertical
Works perfectly fine, also restores the focused cell of the selection :+1:

uklotzde commented on src/library/baseexternalplaylistmodel.h:
```diff
@@ -28,11 +30,19 @@ class BaseExternalPlaylistModel : public BaseSqlTableModel {
     Qt::ItemFlags flags(const QModelIndex &index) const override;
     CapabilitiesFlags getCapabilities() const override;
+    QString modelKey() const override {
+        return pointerToQString(this) %
```
Please use '+' instead of '%' as stated in the Coding Guidelines

> From Mixxx 2.3 we set QT_USE_QSTRINGBUILDER to use QStringBuilder for operator+. Use + in favour of % for better readability.

uklotzde commented on src/library/browse/browsetablemodel.h:
```diff
@@ -72,6 +72,13 @@ class BrowseTableModel final : public QStandardItemModel, public virtual TrackMo
     int columnIndexFromSortColumnId(TrackModel::SortColumnId sortColumn) override;
+    QString modelKey() const override {
+        return QString("0x%1/%2#%3")
+                .arg((quintptr)this, QT_POINTER_SIZE * 2, 16, QChar('0'))
```
More C-style casts, redundant code and no QStringBuilder. I don't repeat review comments multiple times, please always fix all occurrences.

uklotzde replied:
The purpose of reviews is **not** to amend every single line of code and to dictate you what to do. The best feedback you could get from a review are just questions.

I spend the time for a thorough review. Please spend your time to read the feedback carefully to get the most out of it. Try to understand what a reviewer was thinking, consider the suggestions and come up with an even solution that would have prevented all questions, misunderstandings, and concerns in the first place.

poelzi replied:
must have sliped through, thought I changed all modelKey() methodes

poelzi commented on src/library/baseexternalplaylistmodel.cpp:
```diff
@@ -9,16 +9,16 @@
           m_playlistTracksTable(playlistTracksTable),
-          m_trackSource(trackSource) {
+          m_trackSource(trackSource),
+          m_currentPlaylistId(0) {
```
good catch

Holzhaus commented:
Please avoid force-push after review has already started. That makes it harder to see what was changed since last review.

poelzi committed:
Fix crash on model switch, prevent memory leak

poelzi committed:
Use reusable model key names

poelzi committed:
Save and restore header sorting

Use the model settings to store and restore table header state
for each crate/playlist/etc.
Reimplementation of #2046

poelzi commented:
This patchset now also saves and restores the sorting per trackmodel, since it was very easy to morge the #2046 functionality with this one.

poelzi committed:
Remove problematic default modelKey, instead require implementation.

poelzi committed:
Merge remote-tracking branch 'origin/2.3' into fix_selection

poelzi committed:
Remove unused variable

poelzi committed:
Save also vertical scrolling position

daschuer commented on src/widget/wlibrarytableview.cpp:
```diff
@@ -3,22 +3,30 @@
+// number of items to stop purging
+constexpr int kClearModelStatesLowWatermark = 1000;
+// number of items to start purging
+constexpr int kClearModelStatesHighWatermark = 1100;
```
A wartermark is for me a sign, visible if you hold a piece of paper against light.   

How about 
* kModelStateCacheLowThreshold
* kModelStateCacheHighThreshold

daschuer commented on src/widget/wlibrarytableview.h:
```diff
@@ -47,20 +59,12 @@ class WLibraryTableView : public QTableView, public virtual LibraryView {
-    // executed
-    int m_noSearchVScrollBarPos;
+    // saves scrollposition/selection/etc
+    QMap<QString, ModelState*> m_vModelState;
```
Did you consider to use QCache? It is a FIFO Cache and should work exactly for our purpose.  

Please rename to something better like m_modelStateCache. 
We have no other prefix rule than p for pointer in Mixxx.

poelzi replied:
I obviously did not do my research if some proper battery is included. Fixed.

daschuer commented on src/widget/wlibrarytableview.cpp:
```diff
@@ -53,80 +61,128 @@ WLibraryTableView::WLibraryTableView(QWidget* parent,
+    }
+    ModelState* state;
+    if (m_vModelState.contains(key)) {
+        state = m_vModelState[key];
```
This is a double lookup. Please use m_vModelState[key] directly and check the result for null.

daschuer commented on src/widget/wlibrarytableview.cpp:
```diff
@@ -53,80 +61,128 @@ WLibraryTableView::WLibraryTableView(QWidget* parent,
+    if (!m_vModelState.contains(key)) {
+        return;
+    }
+    ModelState* state = m_vModelState[key];
```
A null check is missing.

daschuer commented on src/library/baseplaylistfeature.cpp:
```diff
@@ -156,6 +156,7 @@ void BasePlaylistFeature::activateChild(const QModelIndex& index) {
     //qDebug() << "BasePlaylistFeature::activateChild()" << index;
     int playlistId = playlistIdFromIndex(index);
     if (playlistId != -1) {
+        emit saveModelState();
```
I think we can get rid of the library signal roundtrip if we move emit saveModelState(); into m_pPlaylistTableModel->setTableModel(playlistId); and connect it directly to WLibraryTableView using model() and qobject_cast to TrackModel

poelzi replied:
I tried this but this does not work and I tried it first. Many features like use the same model or do stuff in the background and save and restore must be done at certain points. For example, the crate model changes the crateId for the same model and calls to active itself. At that point the id variable is already changed but the model still has the old data.

daschuer replied:
OK.

daschuer commented on src/widget/wlibrary.cpp:
```diff
@@ -52,14 +52,9 @@ void WLibrary::switchToView(const QString& name) {
     //qDebug() << "WLibrary::switchToView" << name;
 
-    WTrackTableView* ttView = qobject_cast<WTrackTableView*>(
+    LibraryView* oldLibraryView = dynamic_cast<LibraryView*>(
```
qobject_cast should work and is much faster.

poelzi replied:
LibraryView is not a qobject but a mixin. I always try qobject_cast first :)

daschuer replied:
OK

poelzi committed:
Add action for removing Header Settings, only save when changed

poelzi committed:
Use QCache for model states

poelzi committed:
Use QStringLiteral where a QString is constructed

poelzi committed:
Make QCache usage safe against concurrent deletion

poelzi commented on src/library/baseexternalplaylistmodel.h:
```diff
@@ -28,11 +30,18 @@ class BaseExternalPlaylistModel : public BaseSqlTableModel {
     Qt::ItemFlags flags(const QModelIndex &index) const override;
     CapabilitiesFlags getCapabilities() const override;
+    QString modelKey() const override {
+        return QLatin1String("external/") +
```
I'm always confused when to use which ;)

poelzi committed:
Add option to enable/disable state save functionality.

poelzi committed:
Properly sync menu after header state restoration

poelzi commented:
@daschuer I did some small functional addition/change.
The state is now only saved if the user changed the order while using a crate/playlist and the state save checkbox is enabled.
There is also a action to forget the saved state.
I think this solution is now very powerful. If there is no state set, it just keeps the current state when switching.
If there is a state in a crate/playlist/etc, it is restored. It is only saved if the user actually changes something like order, visible fields or the width.

poelzi commented:
An alternative logical behavior which might be better is this:
We use the library as the master setting, if there is no state saved, we use the configuration of the library, or save an anonymous state that is restored. Currently the view on a crate without saved settings is different depending on which crate was visited before. this might or not be useful. It is useful if you want to copy configuration from one setting to another.

poelzi committed:
Emit sorting changed signal when header is restored for updating model

poelzi committed:
Create independent states only on demand

Use a shared state pre feature but have a action to crete/delete
independent states for each feature. This is less confusing to use.

poelzi commented:
I changed to logic since it felt confusing. Now each feature shares a state like before, but on demand a crate/playlist can get a 
independent state. This feels much more natural to use and allows customization on each feature even per playlist/crate.

poelzi committed:
Only trigger resort if headerRestore failed (saves a select())

poelzi committed:
Fix a major connect bug

loadModel method caused more and more connection to the tracksChanged
signals. After each reload, one more select is issued by the next
playlist change

poelzi commented:
I found one major bug not from this patch. f50255a4be59a1747e67f4bf4f351af02d9882fb need to be cherry picked if this does not land.

poelzi committed:
Use checkbox for stating independent state.

poelzi commented:
@xerus2000 please give this branch a test. I think it will fit your needs.
For each crate/playlist/etc you wish i independent state, you need to activate it from the context menu of the table header.
This is table order, which column, size, etc.
Selection and current index is always saved per crate/playlist....

poelzi committed:
Merge branch '2.3' of https://github.com/mixxxdj/mixxx into fix_selection

ywwg commented:
This saves the sort order for all crates, but not for each specific crate.  For instance, I want my Pop crate sorted by bpm, and my House crate sorted by comment.  (New crates could be sorted by the default library sort)

poelzi commented:
@ywwg by default yes. You have to enable each crate you want a individual sorting in the header of the crate "use independant state"
While testing different implementations of this feature, this felt to best behaviour.

poelzi commented:
This allows also to have different  columns or order in that crate

daschuer commented:
The issue with the current implementation in 2.3 is that uses one state for all playlist/crates is, that you loose the context when navigating between different playlists. 
However i guess the majority of users want to change the header configuration for all playlist / crate at once. 

Now, default I still loose context. IMO the issue is worse, because a different track is highlighted. 
For me the sort order is a canonical information together with the slider and the selected track. Can we **always** store the sort order individual? 
It would be just on extra click if the user needs another sort and luckily the selection is already maintained. 

"use independent state" sounds like developer speech. Can we invert it and call it "Apply changes to all playlists" or such? 
  
The individual selection for each search does not work for good for me. 
I would expect that 
* the selection is maintained as long as possible and restored after search. -> already OK  
* If I select a different track during search the selection should be maintained after search -> Not OK. 

This serves the use case to see a found track in the context of other tracks form the same year. 
Currently I need to scroll manually to the found track, which is kind of annoying.

poelzi commented:
pardon me, but nothing is as broken and annoying as the 2.3 branch behavior and typically one of the reasons to have at least one or two swear words on a gig.
You can find different behavior in the history of this branch, I tried like 3-4 different behaviors for some time but was not satisfied at all.
You can't expect everyone to use crates like you do, I probably use crates quite differently then you. In fact, I use crates more like tags and very seldom visit at a gig, sometimes at home. I spend most time in the library and change my search query.
I don't understand what you mean with context. If I'm in a different crate and have some x selected, the chances that song x is also in the next crate I'm selecting are minimal. The context is the crate and this context is preserved, I switch to another crate with totally different set of tracks, I'm in a different context.
Same is with the library, the context is my search query, If i go back to my query, I would like to have the context (selection) of that.

The problem with totally independent states is, that it gets really hard to adjust table columns/order/etc. You would need to change very crate to the way you want it, which is very annoying. Then, ok, lets say you have a reset state action that resets the state to the default one. So, now, which crate is the default state in the first place. Does it reset and next time you leave it, it gets saved as the default ? But when will the independent state saved again ?

You can try this version
715b81daa275064afd7197d90695290995c4d7d4
or this
04244b670004af5d1a895364b01b3f89a70a1dae

I have at least 50 crates and I want them to have the same layout, maybe I will use some different layout for some crates that are special, but for surely, I don't want to configure all of them. If i have some order I'm looking at crates, I want this order to be preserved unless I choose otherwise.

Apply to all feels terrible to me, loosing all customization at once. Currently, I can easily change my column in the shared state and it applies to most. Those with independent state, I can disable/enable the flag and will have the common state to start with.

If you have some better wording, I'm happy to change it.

xerus2000 commented:
scrolling works fine with my F1 rotary, the state stuff I gotta test

xerus2000 commented:
> The individual selection for each search does not work for good for me.
> I would expect that
> 
>     * the selection is maintained as long as possible and restored after search. -> already OK
> 
>     * If I select a different track during search the selection should be maintained after search -> Not OK.
> 
> 
> This serves the use case to see a found track in the context of other tracks form the same year.
> Currently I need to scroll manually to the found track, which is kind of annoying.

This is a legitimate issue I would also like to see addressed.

xerus2000 commented:
I was a bit confused. I thought the column layout and widths also sync between Library and Crates, so I initially thought the feature wasn't working. Now I created a second crate in my test db and it became clear. Yes, this seems useful.

It would be nice to document the current behavior in the top comment, since you tried a lot it is not easy to find out how it is supposed to work without reading through all the conversations, which not every tester is able to ;)

xerus2000 commented:
Oh and while you're at it, could you change the default column width when adding new columns to more than just a sliver?

ywwg commented:
>  "use independant state"

oh I see!  Yeah I think that's good

ywwg commented:
this works great now that I know about the checkbox! lgtm for functionality

ywwg commented:
> Oh and while you're at it, could you change the default column width when adding new columns to more than just a sliver?

or maybe the default should be to auto-size? (like doubleclicking the column border)

ronso0 commented:
> Oh and while you're at it, could you change the default column width when adding new columns to more than just a sliver?

Please let's not extend this PR even further. The columns can be adjusted in another PR.

poelzi commented:
I gave this PR some more love and it now tries to preserve the selection if possible. There is only one thing left I will try to fix today: scrollTo does not work and I'm try to debug why.

Be-ing commented:
Is the description at the top of the PR still accurate for what this PR has become? If not, please update it.

Also, there is a merge conflict.

daschuer commented:
This branch has now conflicts. 

> There is only one thing left I will try to fix today: scrollTo does not work and I'm try to debug why.

What is the current state?

daschuer commented:
> "use independent state" sounds like developer speech. Can we invert it and call it "Apply changes to all playlists" or such?

So we have an individual "use independent state" flag for each playlist. I did not get this during testing. 
So we have a default group that is configured together and individual playlists. 
What happens if you uncheck "use independent state"? Will it adopt the state of the default group. That would be reasonable. 
How about: 
"[ ] follow default template"

poelzi commented:
@daschuer yes, it instantly configures like the default group. So, If added for example a new row or made general layout adjustments, you can do it in the default group and then copy this state to the ones you want special configuration like sorting order by toggle it there and set the new old settings.
I did some general changes in my dev branch because I found some behaviour annoying that I changed. It now tries to save the trackids instead of the selection index when possible. I often use filter like "played:0" in the library and after coming back from autodj, totally different tracks are selected. For the current index, it tries to find the track, but if the track is missing uses the index position so you are at least in the same area.
I think I have now an Idea how to behave when the filters are relaxed and will finish that this weekend.

Holzhaus commented:
Maybe "Use global configuration" ? Default sounds like it's the column config that Mixxx ships with.

poelzi commented:
Why is wording so hard ? :)
It is also not the global config as it is shared per feature.
Maybe "[ ] use shared configuration" ?

daschuer commented:
> Maybe "[ ] use shared configuration" 
Perfect :-)

ronso0 commented:
Is there anything to be done here except fixing the conflict and adjusting to `[  ] use shared configuration ` ?
Would love to get this merged.

uklotzde commented on src/widget/wlibrarytableview.cpp:
```diff
@@ -53,81 +57,127 @@ WLibraryTableView::WLibraryTableView(QWidget* parent,
-
-            delta++;
         }
+        // this wired combination works when switching between models
```
weird?

uklotzde reviewed (approved):
I hope everyone involved has tested it thoroughly during the last weeks!? I feel slightly uncomfortable considering the complexity and amount if changes. On the other hand I don't have strong objections.

ronso0 commented:
If necessary I can also try this again this weekend, hopefully with controller.

poelzi commented:
Not yet, ich will give this today some more love and will give a ping when I say it's ready.

Be-ing commented:
ping @poelzi there is now a merge conflict

Be-ing commented:
Pinging again @poelzi. I would hate for this to be the last PR holding up the 2.3.0 release or have to push it back to 2.4.

poelzi commented:
sorry, I'm very busy in the last weeks and did not find any time to work on this. I have a weighted levensteihn distance now for which I need to write some tests and then use it for the selection logic.

Be-ing commented:
So should we retarget this for 2.4?

ronso0 commented:
> Pinging again @poelzi. I would hate for this to be the last PR holding up the 2.3.0 release or have to push it back to 2.4.

me, too.
@poelzi I'm not sure when we currently _expect_ to release 2.3, but in order to not having to push this valuable PR back to 2.4 due to the lack of thorough testing/tests -- and given the fact that this now contains more features than initially planned -- can you imagine to split of the features added last and focus on the selection improvements?

Holzhaus commented:
There's a merge conflict now.

poelzi commented:
I have some time over eastern at latest to give it some more love. The algorithm I implemented is not the right one anyways, but today I had correct idea I think :)

poelzi committed:
rename use independent state to use shared configuration

poelzi committed:
better selection restoration model

- save TrackIds instead of index if possible
- restore model state only if the search query is less specific
- try to select the previous track if no old state exists
- fix scrollTo by executing in a later loop

poelzi commented:
@ronso0 all done now. I tweaked the algorithms a bit more and it feels quite good now.
@daschuer I fixed the case when a query is less specific. It should work for most cases, but the architecture of the query parser needs a refactoring first so where is a proper tokinization phase seperated from the query construction and this is out of scope here. The special case is an queries like  'crate:"something with spaces"' to 'crate:"something with space"' is not considered a reduces search term. It also only affects the cases where a seperate state is stored.
I will likely fix this in 2.4

poelzi committed:
Merge branch '2.3' of https://github.com/mixxxdj/mixxx into fix_selection

daschuer commented:
Just testing and I am unsure if this is a bug. 
* Go to "crate 1" 
* Uncheck "Preview" while "User shared configuration is checked"
* Preview column disappears
* Go to "crate 2" 
* No "Preview" column 
* Go to Tracks
* Preview column is back -> Unexpected
*  All have  "User shared configuration is checked"

daschuer commented:
Messing with all of this, a "restore defaults" might be handy.

daschuer commented on src/widget/wtracktableview.cpp:
```diff
@@ -836,6 +868,46 @@ void WTrackTableView::setSelectedTracks(const QList<TrackId>& trackIds) {
+    selectRow(idx.row());
+    // we need to emit the scrollTo event after the current loop finishes
+    // otherwise some early check in scrollTo will fail and no scroll happens
+    emit(scrollToCurrent());
```
```suggestion
    emit scrollToCurrent();
```

daschuer commented:
If I search in a crate something that is not there, my original selection is lost afterwards. I have just found this, without a use real case. So it is OK?

daschuer commented:
Yes, there is one: If I have sorted my library by year and scroll to the 80th. Than I want to play "Dear Jessy", but it is not there. Than I want to pick another 80th track and delete my query. Now I need to scroll back to the 80th. It would be nice if we can avoid that.

poelzi committed:
fix clazy warning

poelzi committed:
skip empty search terms in query split

poelzi committed:
Add restore default action in tableview header

poelzi committed:
more sensible fallbacks when restoring track selections

poelzi committed:
fix clazy warning

poelzi commented:
@daschuer I fixed your cases as far as I can see. I added a reset action into the header.
The shared configuration is per subsystem, so there is one for crates, one for history, one for the library. Since some subsystems have different columns, they can't easily be shared.

@ronso0 I think the context menu in the header will need some small fix for alignment since reset is not a checkbox.

poelzi committed:
fix ubuntu 18 compatibility

ronso0 commented:
:steam_locomotive: I'll test this as soon as I can with my controller. wil try to free some time tomorrow night.

poelzi commented:
I have no idea why windows is failing

Holzhaus commented:
> I have no idea why windows is failing

Maybe a just a weird fluke, I restarted the build.

poelzi commented:
this is the 2.4 merge of this patchset, https://github.com/poelzi/mixxx/tree/fix-selection-2.4

poelzi commented:
I tested the reset functionality requested by @daschuer , but it seems to work only partially. I tried to fix it today, but since I have very little understanding in that area of mixxx, I can't get it working as expected. Where does the inital configuration come from in the first place, if the protobuf does not exist yet. loadDefaultHeaderState does not do enough, as it does not restore the order or visibility of a column.

ronso0 commented:
> Messing with all of this, a "restore defaults" might be handy.

I assume the defaults are the initial column configuration after the first start of Mixxx, or after an (database) update respectively?
We don't have "restore defaults" by now and users can screw up the headers. Not having it here is not a regression IMO.

Reminder: document the new feature in the manual before release.

daschuer commented:
While the restore default is quite handy, it is OK to skip this if it turns out to be a hard fix.
Does the track Table share the info with one else? 

I can confirm that the keeping the selected track works now like a charm. That really makes a difference. Thank you.

ronso0 commented:
> Just testing and I am unsure if this is a bug.
> 
>     * Go to "crate 1"
>     * Uncheck "Preview" while "User shared configuration is checked"
>     * Preview column disappears
>     * Go to "crate 2"
>     * No "Preview" column
>     * Go to Tracks
>     * Preview column is back -> Unexpected
>     * All have  "User shared configuration is checked"

So "Use shared configuration" is not ideal, I was also a bit confused.
What about "Use '[feature name]' configuration"? Is the feature name shown in sidebar somehow accesible from the qtableview/qheaderview?

ronso0 commented:
Perfecto, selection testing went well as before!

I expected that selction ranges are also restored, but it's only the last (un)selected (=focused) item.
I guess I remember that from long ago with my selection-fix branch (which didn't work in many other regards..)
Though this is definitely not a blocker.

ronso0 commented:
> I expected that selction ranges are also restored, but it's only the last (un)selected (=focused) item.
> Though this is definitely not a blocker.

...and should not be implemented here IMO (unless it's really trivial) considering we need to keep testing the **current** implemetation before merging to overdue 2.3

daschuer commented:
"Use Crates configuration" is a good idea. 

Does the Tracks feature share with anything?

ronso0 commented:
the subitems 'Hidden' and 'Missing', right? that would matter only in those views.
btw Recording is like Computer, just with a fixed folder

Holzhaus commented:
What is the status here? I didn't follow this closely, but it looks more like a whole new feature than a bugfix. Should it really go into 2.3? Can we pull out the bugfix and merge the other stuff to main? Or are you confident that this doesn't break stuff shortly before the release?

ronso0 commented:
I think @uklotzde already summarized it quite well:

> I'm fine with targeting 2.3 if this branch is tested thoroughly by others **before merging**. @poelzi This also requires your commitment to fix any bugs asap that might be discovered **after merging**!

@poelzi So no matter how much testing this receives, nothing can compensate your availability in case we discover bugs after merge. Worst case: if we merge anyway and no one else can fix related bugs we'd need to revert the entire PR.
Very unfortunate, but correct: in case @poelzi has no time right now to finish this for 2.3 and eventually help fixing regressions -- we should postpone this to 2.4

Otoh: the current state of navigating and jumping in between multiple feature with controllers is broken in 2.3 right now. So that can't get any worse IMO.
Header configs aren't stored individually either in 2.3. Worst case would be screwed up headers after each start, right?

@daschuer @ywwg What do you think?
Which scenarios do we need to test?

Holzhaus commented:
> Otoh: the current state of navigating and jumping in between multiple feature with controllers is broken in 2.3 right now. So that can't get any worse IMO.

I agree. It's really unfortunate that this mixes the bugfix with a new feature, so if we merge this we'd violate feature- and string-freeze yet again. But I think the benefits are probably worth it.

> Header configs aren't stored individually either in 2.3. Worst case would be screwed up headers after each start, right?

As with anything that touches cpp code, the worst case is Mixxx segfaults during a set. But I guess the chance is slim ;-)

@daschuer have all your comments been dealt with? Is there anything left to do here? If I test this, is there anything I should test for in particular?

ronso0 commented:
> It's really unfortunate that this mixes the bugfix with a new feature, so if we merge this we'd violate feature- and string-freeze yet again. But I think the benefits are probably worth it.

unfortunately I'm not able to assess the risks of the new header feature.
which still needs to be polished anyway to be ready for the 2.3 release, yet. 

no reply from @poelzi

due to the model state code it looks unrealistic to me to untangle lib-selection fix // individual header state feature for 2.3

pushing this to 2.3.1
@daschuer Feel free to revert this if you're willing to finish this https://github.com/mixxxdj/mixxx/pull/3063#issuecomment-824077485

Be-ing commented:
@poelzi could you rebase this on main now?

ronso0 commented:
How do we want to proceed here to get this into 2.4?
@daschuer @ywwg @Holzhaus @uklotzde 

**ToDo**
- address review comments

**Nice to have**
- add `Use [feature name] configuration` in crates/playlists that have a different column configuration
- add `Reset to default column configuration`

**1)** @poelzi will you have time to finish it until the beta starts in September?

**2)** anyone else has the time and motivation to finish this?

**3)** we merge as it is and fix issues in beta, as well as add the nice-to-have functions

**4)** As we noted earlier: any improvement of the table selection (even with some bugs) is better than the current state. I didn't wrap my head around the header thing, though I offer to pick & rebase the table selection commits only (and remove all header-related changes from those). During beta anyone can jump in to fix bugs that pop up.
Saving the header state will be harder to re-add later on, though.

What do you think?

ywwg commented:
This PR seems to have gotten a bit out of hand -- it is huge for what it says it is fixing, and it sounds like it causes some known regressions.  I don't feel comfortable merging this in as-is and then hoping someone cleans up the resulting issues.  (I worked very hard on header state code and it would be very discouraging to have a PR merged that knowingly breaks it).

Sometimes things get away from us and I think for the good of the codebase in the long term we need to think very carefully about how to proceed.  Nobody likes to write code that doesn't get used, but this happens all the time in companies, in academia, etc.  We can't let a sunk cost fallacy push us towards the wrong decision. What I would prefer to do here is to raid this PR for the Actual Bug Fixes, and make a clean PR that fixes those issues alone.  Then, we can talk about the design of the new feature.

ronso0 commented:
I've started to pick & rebase commits one by one (trying to merge requested changes ad-hoc) to include only the selection restoration fixes and new 'save model state'.

poelzi commented:
@ronso0 I rebased this branch some time ago at https://github.com/poelzi/mixxx/tree/fix-selection-2.4

ronso0 commented:
marking this as draft.
@ywwg Maybe you want to take a look a the commits saving the table header per view.

ywwg commented:
yeah I'll take a look

ronso0 commented:
Closing this stale PR now since the track view changes have been adopted by now, due to the enormous amount of conflicts and to clean up the list of open PRs.
